porch/func/wrapper-server/main.go

```diff
@@ -98,15 +99,41 @@ func (e *singleFunctionEvaluator) EvaluateFunction(ctx context.Context, req *pb.
 
 	err := cmd.Run()
 	var exitErr *exec.ExitError
-	if err != nil && !errors.As(err, &exitErr) {
-		return nil, status.Errorf(codes.Internal, "Failed to execute function %q: %s (%s)", req.Image, err, stderr.String())
+	outbytes := stdout.Bytes()
+	stderrBytes := stderr.Bytes()
+	if err != nil {
+		if errors.As(err, &exitErr) {
+			// If the exit code is non-zero, we will try to embed the structured results and content from stderr into the error message.
+			results, pe := parseResults(outbytes)
+			if pe != nil {
+				// If we can't parse the output resource list, we only surface the content in stderr.
+				err = fmt.Errorf("KRM function failed with stderr: %v", string(stderrBytes))
+			}
+			err = fmt.Errorf("KRM function failed with structured results: %v and stderr: %v", results.Error(), string(stderrBytes))
+		} else {
+			return nil, status.Errorf(codes.Internal, "Failed to execute function %q: %s (%s)", req.Image, err, stderr.String())
+		}
 	}
 
-	outbytes := stdout.Bytes()
 	klog.Infof("Evaluated %q: stdout length: %d\nstderr:\n%v", req.Image, len(outbytes), stderr.String())
 
 	return &pb.EvaluateFunctionResponse{
 		ResourceList: outbytes,
-		Log:          stderr.Bytes(),
-	}, nil
+		Log:          stderrBytes,
+	}, err
+}
+
+// TODO: replace this with fn.ParseResourceList when github.com/GoogleContainerTools/kpt-functions-sdk/go/fn in go.mod
+// is upgraded to cdc9f0cfea190d80397dc05e2b514b2ba1c89b6c or later.
+func parseResults(resourceList []byte) (fn.Results, error) {
+	rlObj, err := fn.ParseKubeObject(resourceList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse input bytes: %w", err)
```

Comment:
nit: this message may be confusing since we are parsing result and the error will read "failed to parse _input_ bytes"